tests/exporter/test_ingestexportservice.py

```diff
@@ -468,6 +469,60 @@ def test_add_links_no_duplicates(self):
         self.assertTrue(links.get_links()[0] == mock_link)
         self.assertTrue(links.get_links()[1] == another_mock_link)
 
+    SUBMISSION_URL = 'http://api.ingest.data.humancellatlas.org/SubmissionEnvelope/1234'
+
+    def test_upload_error_posted_to_ingest_api(self):
+        with self.assertRaises(Exception):
+            # given:
+            exporter = IngestExporter(ingest_api=self.mock_ingest_api, dss_api=self.mock_dss_api,
+                                      staging_api=self.mock_staging_api)
+            # and:
+            exporter.logger.info = MagicMock()
+            exporter.get_all_process_info = MagicMock()
+            exporter.get_metadata_by_type = MagicMock()
+            exporter.prepare_metadata_files = MagicMock()
+            exporter.bundle_links = MagicMock()
+            exporter.create_bundle_manifest = MagicMock()
+            exporter._extract_submission_url = MagicMock(return_value=self.SUBMISSION_URL)
```

Comment:
My test isn't dependant on the private implantation of getting a submission url. The method I'm testing just needs any submission url to be set to get into the code block I'm actually testing.

Reply:
The problem with this method of testing is that it indicates that the test knows too much about the implementation details of the the system under test. If for some reason the method under test stopped using `_extract_submission_url` in the future, this test will fail but not because the behaviour it's testing for necessarily changed; only because the implementation changed.

Reply:
Agreed. Mocking the submission rather than the _extract_submission_url.